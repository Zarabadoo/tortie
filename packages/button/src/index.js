import TrtButton from "./Button.vue";
export default TrtButton;
