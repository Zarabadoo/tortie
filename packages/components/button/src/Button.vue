<template>
  <button class="trt-button" type="button" v-bind="$attrs" v-on="$listeners">
    <slot />
  </button>
</template>

<script>
export default {
  name: "TrtButton",
  inheritAttrs: false
};
</script>

<style lang="scss" scoped>
.trt-button {
  padding: 0.6rem 1rem;
  font: {
    size: 1rem;
    weight: bold;
  }
  line-height: 1.5;
  vertical-align: baseline;
  color: rgba(#fff, 0.92);
  border: 0.1rem solid darken(blue, 4);
  border-radius: 0.2rem;
  background: darken(blue, 3);
  cursor: pointer;
  transition: all 0.2s ease-out;

  &:hover,
  &:focus {
    &:not([disabled]) {
      color: #fff;
      border-color: blue;
      background: blue;
    }
  }

  &:focus:not([disabled]) {
    outline: none;
    box-shadow: 0 0 0.6rem rgba(blue, 0.8);
  }

  &:active {
    transform: scale(0.98) translateY(0.01rem);
  }

  &[disabled] {
    opacity: 0.4;
    filter: grayscale(60%);
    cursor: not-allowed;
  }

  svg {
    display: inline-block;
    margin: 0 0 -0.2rem 0.2rem;
    fill: currentColor;
    width: 1.2rem;
    height: 1.2rem;
  }
}
</style>
