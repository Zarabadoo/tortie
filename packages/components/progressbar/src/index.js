export { default as TrtProgressbar } from "./Progressbar.vue";
export { default as TrtProgressbarLinear } from "./ProgressbarLinear.vue";
export { default as TrtProgressbarRadial } from "./ProgressbarRadial.vue";
