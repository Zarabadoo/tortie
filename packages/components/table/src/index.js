import TrtTable from "./Table.vue";
export default TrtTable;
