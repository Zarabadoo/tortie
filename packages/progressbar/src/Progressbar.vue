<script>
export default {
  name: "TrtProgressbar",
  data() {
    return {
      progressFactor: null
    };
  },
  props: {
    /**
     * Show that the progress bar is waiting for a process to finish.
     */
    busy: {
      default: false,
      type: Boolean
    },

    /**
     * Set the minimum value available for the progress bar.
     */
    valuemin: {
      default: 0,
      type: Number
    },

    /**
     * Set the maximum value for the progress bar.
     */
    valuemax: {
      default: 100,
      type: Number
    },

    /**
     * Set the value of the progress bar.
     *
     * This should only be provided if there is a known value. If the value is
     * not known, provide `null` as an initial value.
     */
    valuenow: {
      default: null,
      type: Number
    },

    /**
     * Provide an human readable representation of `valuenow`.
     *
     * An example of this would be for shipping progress. Instead of a numeric
     * value, it would be better to provide terms like "Order recieved",
     * "Shipped", and "Delivered".
     */
    valuetext: {
      default: null,
      type: String
    }
  },
  methods: {
    /**
     * Process values and store a factor of progress
     */
    processFactor() {
      let value;
      const now = this.valuenow;
      const min = this.valuemin;
      const max = this.valuemax;

      switch (true) {
        case now === null || now <= min:
          value = 0;
          break;

        case now > max:
          value = 1;
          break;

        default:
          value = (now - min) / (max - min);
          break;
      }

      this.progressFactor = value;
    }
  },
  mounted() {
    this.processFactor();
  },
  watch: {
    valuenow() {
      this.processFactor();
    }
  },
  render() {
    return this.$scopedSlots.default({
      progressFactor: this.progressFactor,
      attributes: {
        role: "progressbar",
        "aria-valuenow": this.valuenow,
        "aria-valuemin": this.valuemin,
        "aria-valuemax": this.valuemax,
        "aria-valuetext": this.valuetext,
        "aria-busy": this.busy,
        "aria-live": "polite"
      }
    });
  }
};
</script>
