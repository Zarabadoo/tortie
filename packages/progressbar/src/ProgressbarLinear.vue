<template>
  <trt-progressbar v-bind="$attrs" v-slot="{ progressFactor, attributes }">
    <div class="trt-progressbar-linear" v-bind="attributes">
      <div class="trt-progressbar-linear__track">
        <div
          class="trt-progressbar-linear__indicator"
          :style="`transform: scaleX(${progressFactor})`"
        ></div>
      </div>
    </div>
  </trt-progressbar>
</template>

<script>
import TrtProgressbar from "./Progressbar.vue";

export default {
  name: "TrtProgressbarLinear",
  components: {
    TrtProgressbar
  }
};
</script>

<style lang="scss" scoped>
.trt-progressbar-linear {
  margin: 0 1rem;

  &__track {
    display: flex;
    justify-content: flex-start;
    align-items: stretch;
    width: 100%;
    height: 1rem;
    overflow: hidden;
    border-radius: 0.2rem;
    background: rgba(#000, 0.14);
    box-shadow: inset 0 0 0.6rem rgba(blue, 0.4);
  }

  &__indicator {
    flex: 1 1 100%;
    transform-origin: left;
    background: blue;
  }
}
</style>
