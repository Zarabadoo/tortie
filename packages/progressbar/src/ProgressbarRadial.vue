<template>
  <trt-progressbar v-bind="$attrs" v-slot="{ progressFactor, attributes }">
    <div class="trt-progressbar-radial" v-bind="attributes">
      <svg
        role="presentation"
        :viewBox="viewBox"
        class="trt-progressbar-radial__progress"
      >
        <circle
          ref="track"
          class="trt-progressbar-radial__track"
          v-bind="circleAttrs"
        ></circle>

        <circle
          ref="value"
          class="trt-progressbar-radial__indicator"
          v-bind="circleAttrs"
          :style="circleStyles(progressFactor)"
        ></circle>
      </svg>
    </div>
  </trt-progressbar>
</template>

<script>
import TrtProgressbar from "./Progressbar.vue";

export default {
  name: "TrtProgressbarRadial",
  components: {
    TrtProgressbar
  },
  data() {
    return {
      diameter: 0,
      center: 0,
      radius: 100 / (Math.PI * 2) // Make the offset easily divisible by 100
    };
  },
  computed: {
    /**
     * Attributes for the SVG circles.
     *
     * @returns {object} Attribute values to be bound to circle elements
     */
    circleAttrs() {
      return {
        cx: this.center,
        cy: this.center,
        r: this.radius
      };
    },

    /**
     * Values for the SVG viewbox.
     *
     * @returns {string} Value for the SVG `viewBox` property
     */
    viewBox() {
      return this.diameter ? `0 0 ${this.diameter} ${this.diameter}` : null;
    }
  },
  methods: {
    /**
     * Calculates and stores the diameter and center of the SVG.
     *
     * Takes into account the radius size along with computed styles for the
     * stroke width.
     */
    calculateDiameter() {
      const trackStroke = parseFloat(
        getComputedStyle(this.$refs.track).strokeWidth
      );
      const valueStroke = parseFloat(
        getComputedStyle(this.$refs.value).strokeWidth
      );

      let stroke = valueStroke > trackStroke ? valueStroke : trackStroke;

      this.diameter = this.radius * 2 + stroke;
      this.center = this.diameter / 2;
    },

    /**
     * Calculates the styles to represent the progress along the circle path.
     *
     * Of note is the use of `styles` and not a direct attribute on the
     * element. This is done since these inline styles will take priority over
     * any stylesheet-based styles used for a pure CSS implementation as these
     * will be more accurate.
     *
     * @return {object} A set of attributes to bind to the progress circle
     */
    circleStyles(ratio) {
      return {
        strokeDasharray: `${100 * ratio} 100`
      };
    }
  },
  mounted() {
    this.calculateDiameter();
  }
};
</script>

<style lang="scss" scoped>
.trt-progressbar-radial {
  width: 10rem;
  margin: 0 1rem;

  &__progress {
    max-width: 100%;
    transform: rotate(-90deg);
  }

  &__track,
  &__indicator {
    fill: none;
    stroke-width: 2.5;
  }

  &__track {
    stroke: rgba(#000, 0.14);
  }

  &__indicator {
    stroke: blue;
  }
}
</style>
