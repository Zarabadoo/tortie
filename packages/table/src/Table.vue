<template>
  <table class="trt-table">
    <caption
      class="trt-table__caption"
      v-if="!!$slots.default || !!$scopedSlots.default"
    >
      <slot></slot>
    </caption>

    <thead class="trt-table__thead">
      <tr class="trt-table__row">
        <th
          class="trt-table__head"
          v-for="(header, h) in processedHeaders"
          :key="h"
        >
          <slot :name="`header-${header.id}`" :header="header">
            <slot name="header" :header="header">
              {{ header.label }}
            </slot>
          </slot>
        </th>
      </tr>
    </thead>

    <tbody class="trt-table__body">
      <tr class="trt-table__row" v-for="(row, r) in rows" :key="r">
        <td
          class="trt-table__cell"
          v-for="(header, c) in processedHeaders"
          :key="c"
        >
          <slot
            :name="`cell-${header.id}`"
            :cell="getColumn(header.id, row)"
            :row="row"
          >
            <slot name="cell" :cell="getColumn(header.id, row)" :row="row">
              {{ getColumn(header.id, row).value }}
            </slot>
          </slot>
        </td>
      </tr>
    </tbody>

    <tfoot v-if="!!footers" class="trt-table__foot">
      <tr class="trt-table__row">
        <td class="trt-table__cell">Foot</td>
      </tr>
    </tfoot>
  </table>
</template>

<script>
export default {
  name: "TrtTable",
  props: {
    footers: {
      type: Array,
      default: null
    },
    headers: {
      type: Array,
      default: () => []
    },
    rows: {
      type: Array,
      default: () => []
    }
  },
  computed: {
    processedHeaders() {
      const columns = [].concat(
        ...this.rows.map(row => {
          return row.columns.map(column => {
            return column.column;
          });
        })
      );

      return columns
        .filter((column, i) => {
          return columns.indexOf(column) === i;
        })
        .map((id, index) => {
          const config = this.headers.filter(h => {
            return h.id === id;
          });

          if (config.length === 1) {
            config[0].weight = config[0].weight || index;
            return config[0];
          }

          return {
            id: id,
            label: id,
            weight: index
          };
        })
        .sort((a, b) => {
          return a.weight - b.weight;
        });
    }
  },
  methods: {
    getColumn(id, row) {
      return row.columns.filter(column => {
        return column.column === id;
      })[0];
    }
  }
};
</script>

<style lang="scss">
.trt-table {
  $_parent: &;

  width: 100%;
  margin: 1rem 0;
  border-collapse: collapse;
  font-family: sans-serif;
  color: rgba(#000, 0.88);

  &__row {
    border-bottom: 1px solid #888;

    &:first-child {
      border-top: 1px solid #888;
    }
  }

  &__head,
  &__cell {
    padding: 0.3rem;
    text-align: left;

    + #{$_parent}__head,
    + #{$_parent}__cell,
    &:first-child {
      border-left: 1px solid #888;
    }

    &:last-child {
      border-right: 1px solid #888;
    }
  }

  &__head {
    vertical-align: bottom;
    color: rgba(#fff, 0.88);
    font-weight: bold;
    background: rgba(#000, 0.4);
  }

  &__cell {
    vertical-align: top;
    background: rgba(#fff, 0.8);
  }

  &__caption {
    padding: 0.3rem;
    font: {
      size: 0.8rem;
      style: italic;
    }
  }
}
</style>
