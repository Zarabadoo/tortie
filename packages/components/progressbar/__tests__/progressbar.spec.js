/* global describe, it, expect */

import { shallowMount } from "@vue/test-utils";
import { TrtProgressbar, TrtProgressbarLinear } from "../src";

describe("Progressbar.vue", () => {
  it("renders slot", () => {
    const msg = "new message";
    const wrapper = shallowMount(TrtProgressbar, {
      slots: {
        default: [msg]
      },
      propData: {
        valuenow: "40"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("ProgressbarLinear.vue", () => {
  it("renders slot", () => {
    const msg = "new message";
    const wrapper = shallowMount(TrtProgressbarLinear, {
      slots: {
        default: [msg]
      },
      propData: {
        valuenow: "40"
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
