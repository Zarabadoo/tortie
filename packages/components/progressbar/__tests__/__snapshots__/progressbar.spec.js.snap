// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progressbar.vue renders slot 1`] = `undefined`;

exports[`ProgressbarLinear.vue renders slot 1`] = `<trt-progressbar-stub valuemin="0" valuemax="100"></trt-progressbar-stub>`;
